psydac/ddm/cart.py

```diff
@@ -310,6 +316,40 @@ def get_shift_info( self, direction, disp ):
 
         return self._shift_info[ direction, disp ]
 
+    #---------------------------------------------------------------------------
+    def get_shared_memory_subdivision( self, shape ):
+
+        assert len(shape) == self._ndims
+
+        try:
+            nthreads , block_shape = compute_dims( self._num_threads, shape , [2*p for p in self._pads])
+        except ValueError:
+            print("Cannot compute dimensions with given input values!")
+            self.comm.Abort(1)
+
+        # Know coords of all threads
+        coords_from_rank = np.array([np.unravel_index(rank, nthreads) for rank in range(self._num_threads)])
+        rank_from_coords = np.zeros([n+1 for n in nthreads], dtype=int)
+        for r in range(self._num_threads):
+            c = coords_from_rank[r]
+            rank_from_coords[tuple(c)] = r
+
+        for i in range(self._ndims):
+            ind = [slice(None,None)]*self._ndims
+            ind[i] = nthreads[i]
+            rank_from_coords[tuple(ind)] = self._num_threads
```

Comment:
Why is loop this needed?

Reply:
in the variable `rank_from_coords` we store the rank of each thread according to its coordinates, we need it to have easy access to neighboring threads, you just need to move `+1` or `-1` in the thread coordinates in the direction that you want, the loop that you mentioned helps avoid the corner case where a thread is on the boundary, it does not have a neighbor in  some directions, so we provide an extra rank that indicates that we are out of the domain's boundary

Reply:
This solution sounds strange to me. In order to know if a thread is on the domain's boundary, you just need to check its coordinates: if `coords[i] == 0` or `coords[i] == nthreads[i]-1` you know that you are on the boundary along direction `i`.

Reply:
yes, but I don't want to check that every time, otherwise I will have ifs in the code

Reply:
You already need if statements to treat the case where `coords[i] == 0`, don't you?

Reply:
no, I don't

Reply:
```python
def assembly_linear_form(global_test_basis_v_1, global_test_basis_v_2, 
                        global_span_v_1, global_span_v_2, global_x1, global_x2, 
                        test_v_p1, test_v_p2, n_element_1, n_element_2, k1, k2, pad1, pad2, 
                        g_vec_v_0_aykkxp65, coords_from_rank, rank_from_coords, global_thread_starts_1, global_thread_starts_2,
                        global_thread_ends_1, global_thread_ends_2, num_threads):

    from numpy import array, zeros, zeros_like, floor, sin, pi
    from pyccel.stdlib.internal.openmp import omp_get_thread_num
    local_x1 = zeros_like(global_x1[0, : ])
    local_x2 = zeros_like(global_x2[0, : ])

    #$ omp parallel default(private) &
    #$ shared(coords_from_rank,rank_from_coords,global_thread_starts_1,global_thread_starts_2,&
    #$ global_thread_ends_1,global_thread_ends_2,global_test_basis_v_1,global_test_basis_v_2,global_span_v_1,&
    #$ global_span_v_2,global_span_v_3,global_x1,global_x2,global_x3,g_vec_v_0_aykkxp65) &
    #$ firstprivate(test_v_p1,test_v_p2,n_element_1,n_element_2,k1,k2,pad1,pad2,num_threads)

    l_vec_v_0_aykkxp65 = zeros((5, 5), dtype='float')

    thread_id = omp_get_thread_num()
    thread_coords_1 = coords_from_rank[thread_id,0]
    thread_coords_2 = coords_from_rank[thread_id,1]

    global_thread_size_1 = 1 + global_thread_ends_1[thread_coords_1] - global_thread_starts_1[thread_coords_1]
    global_thread_size_2 = 1 + global_thread_ends_2[thread_coords_2] - global_thread_starts_2[thread_coords_2]

    local_thread_starts_1 = array((global_thread_starts_1[thread_coords_1], int((1/2)*global_thread_size_1) + global_thread_starts_1[thread_coords_1]))
    local_thread_starts_2 = array((global_thread_starts_2[thread_coords_2], int((1/2)*global_thread_size_2) + global_thread_starts_2[thread_coords_2]))

    local_thread_ends_1 = array((int((1/2)*global_thread_size_1) + global_thread_starts_1[thread_coords_1], global_thread_ends_1[thread_coords_1]+1))
    local_thread_ends_2 = array((int((1/2)*global_thread_size_2) + global_thread_starts_2[thread_coords_2], global_thread_ends_2[thread_coords_2]+1))

    for local_i_element_1 in range(0, 2, 1):
        for local_i_element_2 in range(0, 2, 1):
            #$omp barrier
            for i_element_1 in range(local_thread_starts_1[local_i_element_1], local_thread_ends_1[local_i_element_1], 1):
                local_x1[ : ] = global_x1[i_element_1, : ]
                span_v_1 = global_span_v_1[i_element_1]
                for i_element_2 in range(local_thread_starts_2[local_i_element_2], local_thread_ends_2[local_i_element_2], 1):
                    local_x2[ : ] = global_x2[i_element_2, : ]
                    span_v_2 = global_span_v_2[i_element_2]
                    for i_basis_1 in range(0, 5, 1):
                        for i_basis_2 in range(0, 5, 1):
                            contribution_v_0_aykkxp65 = 0.0
                            for i_quad_1 in range(0, 5, 1):
                                x1 = local_x1[i_quad_1]
                                v_0_1 = global_test_basis_v_1[i_element_1,i_basis_1,0,i_quad_1]
                                v_0_1_x1 = global_test_basis_v_1[i_element_1,i_basis_1,1,i_quad_1]
                                v_1_1 = global_test_basis_v_1[i_element_1,i_basis_1,0,i_quad_1]
                                v_1_1_x1 = global_test_basis_v_1[i_element_1,i_basis_1,1,i_quad_1]
                                v_2_1 = global_test_basis_v_1[i_element_1,i_basis_1,0,i_quad_1]
                                v_2_1_x1 = global_test_basis_v_1[i_element_1,i_basis_1,1,i_quad_1]
                                for i_quad_2 in range(0, 5, 1):
                                    x2 = local_x2[i_quad_2]
                                    v_0_2 = global_test_basis_v_2[i_element_2,i_basis_2,0,i_quad_2]
                                    v_0_2_x2 = global_test_basis_v_2[i_element_2,i_basis_2,1,i_quad_2]
                                    v_1_2 = global_test_basis_v_2[i_element_2,i_basis_2,0,i_quad_2]
                                    v_1_2_x2 = global_test_basis_v_2[i_element_2,i_basis_2,1,i_quad_2]
                                    v_2_2 = global_test_basis_v_2[i_element_2,i_basis_2,0,i_quad_2]
                                    v_2_2_x2 = global_test_basis_v_2[i_element_2,i_basis_2,1,i_quad_2]
                                    v_0 = v_0_1*v_0_2
                                    v_1 = v_1_1*v_1_2
                                    v_0_x2 = v_0_1*v_0_2_x2
                                    v_1_x2 = v_1_1*v_1_2_x2
                                    v_2_x2 = v_2_1*v_2_2_x2
                                    v_0_x1 = v_0_1_x1*v_0_2
                                    v_1_x1 = v_1_1_x1*v_1_2
                                    v_2_x1 = v_2_1_x1*v_2_2
                                    contribution_v_0_aykkxp65 += 2*pi**2*v_0*sin(pi*x1)*sin(pi*x2)

                            l_vec_v_0_aykkxp65[i_basis_1,i_basis_2] = contribution_v_0_aykkxp65
                    g_vec_v_0_aykkxp65[pad1 + span_v_1 - test_v_p1 : 1 + pad1 + span_v_1,pad2 + span_v_2 - test_v_p2 : 1 + pad2 + span_v_2] += l_vec_v_0_aykkxp65[ : , :]
                
    #$ omp end parallel
    return
    ```

Reply:
this is an example of the generated code for the linear form, this is what is generated right now, we have another version that is commented that uses `rank_from_coords`, the second version is faster  because it doesn't have a global barrier, but needs some work and checks

Reply:
OK. Please add comments in the code to help other developers

Reply:
done

Reply:
where?

Reply:
in this [fe9a2ba](https://github.com/pyccel/psydac/pull/156/commits/fe9a2ba246b19895245ed84abe30955202293802)